### PR TITLE
feat: Support Amplify Android Lazy Types and Custom Selection Set

### PR DESCRIPTION
**Pull request by tylerjroach**
#### Description of changes
Amplify Android will now support lazy loading and custom selection set with the existing cli.json `generatemodelsforlazyloadandcustomselectionset` flag already in use by iOS.

#### Issue #, if available
Amplify Android Work: https://github.com/aws-amplify/amplify-android/tree/tjroach/custom-selection-set
Documentation: TBD

#### Description of how you validated changes
The newly generated models have been tested in an Amplify Android feature branch.

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-codegen/blob/main/CONTRIBUTING.md#tests)
- [ ] Relevant documentation is changed or added (and PR referenced)
- [x] Breaking changes to existing customers are released behind a feature flag or major version update
- [x] Changes are tested using sample applications for all relevant platforms (iOS/android/flutter/Javascript) that use the feature added/modified

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by tylerjroach** [comment]
Closing until the above comment is resolved.

File: packages/appsync-modelgen-plugin/src/visitors/appsync-java-visitor.ts [hunk 1]
```diff
@@ -494,9 +532,12 @@ export class AppSyncModelJavaVisitor<
       const fieldName = this.getFieldName(field);
       const methodName = this.getStepFunctionName(field);
       const returnType = this.getStepInterfaceName('Build');
-      const argumentType = this.getNativeType(field);
+      const argumentType = this.getNativeType(field, true);
       const argumentName = this.getStepFunctionArgumentName(field);
-      const body = [`this.${fieldName} = ${argumentName};`, `return this;`].join('\n');
+      const assignment = this.isModelReference(field) ?
+      `this.${fieldName} = new LoadedModelReferenceImpl<>(${argumentName});` :
+      `this.${fieldName} = ${argumentName};`
```
Comment (lawmicha) [inline packages/appsync-modelgen-plugin/src/visitors/appsync-java-visitor.ts:539]: nit: add indentation, I see similar ternary conditional operator used at line 515/516 with indentation

File: packages/appsync-modelgen-plugin/src/languages/java-declaration-block.ts [hunk 2]
```diff
@@ -3,7 +3,7 @@ import { StringValueNode, NameNode } from 'graphql';
 import stripIndent from 'strip-indent';
 
 // Todo: PR to @graphql-codegen/java-common to support method exceptions and comment
-export type Access = 'private' | 'public' | 'protected';
+export type Access = 'private' | 'public' | 'protected' | '';
```
[the remark this answers is not in the document]
Comment (tylerjroach) [reply packages/appsync-modelgen-plugin/src/languages/java-declaration-block.ts:6]: It is used here:
https://github.com/aws-amplify/amplify-codegen/pull/699/files#diff-8fdb1b5aed85eca745295520073edb2af47d27fdf8cff3173487f147b11e5fdaR1205

[the remark this answers is not in the document]
Comment (lawmicha) [reply packages/appsync-modelgen-plugin/src/languages/java-declaration-block.ts:6]: ah got it, i'm guessing this is to take the default Access, private, when '' is used?

[the remark this answers is not in the document]
Comment (tylerjroach) [reply packages/appsync-modelgen-plugin/src/languages/java-declaration-block.ts:6]: Not quite. Adding no modifier makes it `package-private` in Java, which is different than both private and protected.

File: packages/appsync-modelgen-plugin/src/visitors/appsync-java-visitor.ts [hunk 3]
```diff
@@ -33,15 +36,22 @@ export class AppSyncModelJavaVisitor<
...
     // This flag is going to be used to tight-trigger on JS implementations only.
     const shouldImputeKeyForUniDirectionalHasMany = false;
     this.processDirectives(shouldUseModelNameFieldInHasManyAndBelongsTo, shouldImputeKeyForUniDirectionalHasMany);
 
-    if (this._parsedConfig.generate === 'loader') {
+    if (this._parsedConfig.generate === CodeGenGenerateEnum.loader) {
       return this.generateClassLoader();
+    } else if (this._parsedConfig.generate === CodeGenGenerateEnum.metadata) {
+      return this.generateModelPathClasses();
     }
```
Comment (lawmicha) [inline packages/appsync-modelgen-plugin/src/visitors/appsync-java-visitor.ts:55]: I was curious when are these used? It looks like the tests are using `CodeGenGenerateEnum.code` so I'm assuming that's the default for `amplify codegen models` from an Android project. How does the developer choose to generate "loader" or "metadata" over "code"?

File: packages/appsync-modelgen-plugin/src/visitors/appsync-java-visitor.ts [hunk 4]
```diff
@@ -33,15 +36,22 @@ export class AppSyncModelJavaVisitor<
   protected additionalPackages: Set<string> = new Set();
   protected usingAuth: boolean = false;
 
+  // We override lazy generation to also only generate lazy models if data store is disabled
+  protected override isGenerateModelsForLazyLoadAndCustomSelectionSet(): boolean {
+    return super.isGenerateModelsForLazyLoadAndCustomSelectionSet() && !(this.config.isDataStoreEnabled ?? false);
+  }
```
Comment (lawmicha) [inline packages/appsync-modelgen-plugin/src/visitors/appsync-java-visitor.ts:42]: nit: maybe `APIOnly` somewhere in the naming of this computed property? `isGenerateModelsForLazyLoadAndCustomSelectionSetAPIOnly` or `isGenerateAPIModelsForLazyLoadAndCustomSelectionSet`

File: packages/appsync-modelgen-plugin/src/preset.ts [hunk 5]
```diff
@@ -28,16 +29,22 @@ export type AppSyncModelCodeGenPresetConfig = {
...
   models: TypeDefinitionNode[],
+  manyToManyJoinModels: TypeDefinitionNode[],
 ): Types.GenerateOptions[] => {
   const config: Types.GenerateOptions[] = [];
   const modelFolder = options.config.overrideOutputDir
     ? [options.config.overrideOutputDir]
     : [options.baseOutputDir, ...GENERATED_PACKAGE_NAME.split('.')];
+
+  // Only generate lazy models if feature flag enabled and datastore is not being used.
+  const generateLazyModels = options.config.generateModelsForLazyLoadAndCustomSelectionSet && !options.config.isDataStoreEnabled
```
Comment (lawmicha) [inline packages/appsync-modelgen-plugin/src/preset.ts:46]: nit: `generateLazyModelsForAPIOnly`, `generateAPILazyModels`

File: packages/appsync-modelgen-plugin/src/preset.ts [hunk 6]
```diff
@@ -4,6 +4,7 @@ import { join } from 'path';
 import { JAVA_SCALAR_MAP, SWIFT_SCALAR_MAP, TYPESCRIPT_SCALAR_MAP, DART_SCALAR_MAP, METADATA_SCALAR_MAP } from './scalars';
 import { LOADER_CLASS_NAME, GENERATED_PACKAGE_NAME } from './configs/java-config';
 import { graphqlName, toUpper } from 'graphql-transformer-common';
+import { ObjectTypeDefinitionNode, StringValueNode } from "graphql/language/ast";
```
Comment (lawmicha) [inline packages/appsync-modelgen-plugin/src/preset.ts:7]: Forgot to remove these in the refactor (https://github.com/aws-amplify/amplify-codegen/pull/725). I believe they are no longer used.
```suggestion
```

File: packages/appsync-modelgen-plugin/src/visitors/appsync-java-visitor.ts (hunk 4 above)
Comment (tylerjroach) [reply packages/appsync-modelgen-plugin/src/visitors/appsync-java-visitor.ts:42]: Problem in that is that this is an override.

File: packages/appsync-modelgen-plugin/src/visitors/appsync-java-visitor.ts (hunk 3 above)
Comment (tylerjroach) [reply packages/appsync-modelgen-plugin/src/visitors/appsync-java-visitor.ts:55]: It comes from preset.ts `generate: 'metadata'`

Comment (lawmicha) [reply packages/appsync-modelgen-plugin/src/visitors/appsync-java-visitor.ts:55]: Oh i see it now, thanks for pointing that out

**Review by lawmicha — DISMISSED** [review]

**Comment by dpilch** [comment]
I started E2E tests for this change: https://us-east-1.console.aws.amazon.com/codesuite/codebuild/594813022831/projects/amplify-codegen-e2e-workflow/batch/amplify-codegen-e2e-workflow:c2e5ffd2-93b3-4422-99b1-eb68b5b43f8f?region=us-east-1

File: packages/appsync-modelgen-plugin/src/preset.ts [hunk 7]
```diff
@@ -49,17 +55,46 @@ const generateJavaPreset = (
         selectedType: modelName,
       },
     });
-  });
```
Comment (dpilch) [inline packages/appsync-modelgen-plugin/src/preset.ts:52]: Is moving the below into the forEach loop intentional?

Comment (tylerjroach) [reply packages/appsync-modelgen-plugin/src/preset.ts:52]: No, great catch. I also removed manyToManyJoinModels from the for loop

**Review by lawmicha — DISMISSED** [review]

**Review by dpilch — DISMISSED** [review]

**Review by dpilch — APPROVED** [review]

**Review by lawmicha — APPROVED** [review]

File: packages/amplify-codegen/src/commands/models.js [hunk 8]
```diff
@@ -260,6 +261,7 @@ async function generateModels(context, generateOptions = null) {
   const generatedCode = await generateModelsHelper({
     schema: loadSchema(apiResourcePath),
     directives: await getDirectives(context, apiResourcePath),
+    isDataStoreEnabled: await isDataStoreEnabled(apiResourcePath),
```
[the remark this answers is not in the document]
Comment (alharris-at) [reply packages/amplify-codegen/src/commands/models.js:264]: Nvm, understood after reviewing further.

**Review by alharris-at — APPROVED** [review]

**Pull request merged**